Gradient palettes get stored in RAM on ESP8266

Hello!
I've noticed that if I use the following syntax to define a palette:

```
DEFINE_GRADIENT_PALETTE( heatmap_gp ) {
  0,     0,  0,  0,   //black
128,   255,  0,  0,   //red
224,   255,255,  0,   //bright yellow
255,   255,255,255 }; //full white
```

the palette gets stored in RAM, not PROGMEM, even though the docs say so.

In `platforms/esp/8266/led_sysdefs_esp8266.h` I found this definition:

```
// Default to NOT using PROGMEM here
#ifndef FASTLED_USE_PROGMEM
# define FASTLED_USE_PROGMEM 0
#endif
```

I have some ~2kB worth of palette definitions in my application and would like to free up some memory.
Can someone explain the issue of enabling PROGMEM on ESP8266?

Researched this properly. **The premise is no longer true — gradient palettes are already in flash on ESP8266.** But the reason is not that anyone fixed `FASTLED_USE_PROGMEM`; it's that the flag stopped being consulted on this platform, and the comment you found is now actively misleading. Details, with evidence.

## 1. Measured: palettes are in flash today

Built `examples/Fire2023` (which uses a real `DEFINE_GRADIENT_PALETTE(hot_gp)`) and `examples/ColorPalette` for `esp8266`, then read the symbols out of the ELF:

```
40239ef0  00000040  T  myRedWhiteBluePalette_p     (ColorPalette, TProgmemPalette16)
4023ac68  00000018  T  hot_gp                      (Fire2023, DEFINE_GRADIENT_PALETTE)
```

Section map from the same binary:

| section | VMA | region |
|---|---|---|
| `.data` | `0x3ffe8000` | DRAM |
| `.rodata` | `0x3ffe8620` | DRAM — **940 bytes total** |
| `.bss` | `0x3ffe89d0` | DRAM |
| `.irom0.text` | `0x40201010`, 242 KB | **flash (XIP)** |

`0x402xxxxx` is the flash-mapped region. Both palettes are in `.irom0.text`. Note also that RAM-resident `.rodata` is only 940 bytes in the whole image — essentially nothing `const` is in RAM.

So your ~2 KB of palette definitions are not costing you RAM on a current FastLED.

## 2. Why — the flag is bypassed on ESP8266

`src/fastled_progmem.h:6-12` dispatches on **platform**, before the flag is ever tested:

```c
#if defined(__EMSCRIPTEN__) || defined(FASTLED_TESTING) || defined(FASTLED_STUB_IMPL)
#include "platforms/null_progmem.h"
#elif defined(ESP8266)
#include "platforms/esp/8266/progmem_esp8266.h"   // <-- ESP8266 exits here
#elif defined(FL_IS_STM32)
...
#else
    ...
    #if (FASTLED_USE_PROGMEM == 1)                // <-- line 45, never reached on ESP8266
```

`progmem_esp8266.h:13-17` then defines `FL_PROGMEM` as real `PROGMEM` and `FL_PGM_READ_*` as real `pgm_read_*`. So ESP8266 gets genuine flash residency **regardless of `FASTLED_USE_PROGMEM` being 0**.

That dispatch was added in `e8ae4f8aeb` / `ba65d939d8` (Aug 2025), six years after you filed this.

## 3. So what's actually still wrong

The comment at `src/platforms/esp/8266/led_sysdefs_esp8266.h:22-25`:

```c
// Default to NOT using PROGMEM here
#ifndef FASTLED_USE_PROGMEM
# define FASTLED_USE_PROGMEM 0
#endif
```

is **stale**. It documents behavior the platform no longer has, and it's what sent you (reasonably) down this path. That's the remaining defect here — a documentation bug, not a memory one.

There's a second wrinkle: setting the flag to 1 is now a **hard compile error**, which directly answers your original question ("Can someone explain the issue of enabling PROGMEM on ESP8266?"). I tried it:

```
$ bash compile esp8266 --examples ColorPalette --defines FASTLED_USE_PROGMEM=1
platforms/esp/compile_test.hpp:13:2: error: #error "FASTLED_USE_PROGMEM should be 0 for ESP8266"
```

So the library now blocks a flag that does nothing. ESP32 carries the identical assert (`compile_test.hpp:59-60`).

## 4. Your 2019 crash

`LoadStoreAlignmentCause` inside `CRGBPalette16::operator=` was an alignment fault: gradient palettes are read with `FL_PGM_READ_DWORD_NEAR` (32-bit loads), and unaligned flash dword reads fault on Xtensa.

That's handled now. `DEFINE_GRADIENT_PALETTE` emits `FL_ALIGN_PROGMEM(4)` before the array (`src/colorutils.h:122-124`), and ESP8266 caps `FL_ALIGN_PROGMEM` at 4 bytes (`progmem_esp8266.h:19-23`) because the platform lacks `memalign`.

Your instinct that it was environment-specific was also right — you noted it reproduced in your full app but not in a trimmed `ColorPalette.ino`. Alignment bugs behave exactly like that, since whether a given array lands aligned depends on what the linker placed before it.

## 5. The ISR hazard people usually assume — checked, doesn't apply

The obvious reason to keep constants out of flash on ESP8266 is that flash reads fault when the cache is disabled. I checked whether FastLED's output path is exposed:

- `clockless_esp8266.h:109` and `clockless_block_esp8266.h:132` are both `FL_IRAM showRGBInternal`, and the `os_intr_lock()` regions (`:122-125`, `:142`) touch only `loadAndScale*` / `stepDithering` / `FastPin` register writes.
- Those primitives are pure RAM arithmetic — `pixel_controller.h:418` reads the RAM pixel buffer, `:431` is `qadd8`, `:445` is `scale8`. `pixel_controller.h` contains **zero** `FL_PGM_*` uses.
- Gamma/`five_bit_hd_gamma` LUT reads exist but are on the **SPI** chipset paths (APA102/HD108), outside any interrupt lock. `Gamma8Impl` even copies its LUT into RAM at construction (`ease.cpp.hpp:411-415`).

So no PROGMEM read happens during clockless output. The hazard isn't why the flag was 0.

## 6. Why it was 0 in the first place

No commit ever explained it. It arrives fully formed in `55690d6832` (2016-01-23, Daniel Garcia, *"ESP8266 first light"*) — the file is created already containing that exact comment and value. Most plausibly the 2016-era core's `.irom`/`pgm_read_*` handling was unreliable and the porter simply left constants in RAM. Whatever the reason, it was superseded in 2025.

## 7. ESP8266 isn't unusual

Only AVR and the Teensy family default to 1. ESP32, RP2040, Apollo3, SAMD21/51, SAM, nRF51/52, MGM240, LPC, WASM and stub all default to 0 — every memory-mapped-flash platform does. What's unusual about ESP8266 is only that its *actual behavior* now contradicts its own flag.

---

## Recommendation

Leaving this open, retitled in spirit as a docs/cleanup item rather than a memory bug. The tractable work is:

1. Replace the misleading comment at `led_sysdefs_esp8266.h:22` with a note that ESP8266 uses PROGMEM unconditionally via the platform dispatch in `fastled_progmem.h:8-9`, and that the flag is vestigial there.
2. Decide what `compile_test.hpp:12-14` should assert now that the flag it guards is inert.
3. `src/platforms/esp/8266/README.md:19,23` needs the same correction. (`src/platforms/readme.md:365-366` already says "ESP8266: PROGMEM supported", which is the accurate statement.)

@Aircoookie — six years late, but: your reading of the code was correct at the time, the crash you hit was a real alignment bug, and the behavior you wanted is what ships today.

**Unrelated find worth its own issue:** the five console fonts (`src/fl/font/console_font_*.h:6-8`, ~12 KB) are declared `FL_PROGMEM` but have no `FL_PGM_*` reader and no consumer anywhere. On ESP8266 — where `FL_PROGMEM` is genuine flash — dereferencing them directly would fault.

--- a/src/platforms/esp/8266/README.md
+++ b/src/platforms/esp/8266/README.md
@@ -17,10 +17,17 @@ ESP8266 family support (Xtensa).
 
 Notes:
 - Typical settings: `FASTLED_USE_PROGMEM=0`, `FASTLED_ALLOW_INTERRUPTS=1`; long ISRs will impact signal quality.
+- `FASTLED_USE_PROGMEM=0` does **not** mean constants live in RAM here — see below.
 
 ## Optional feature defines
 
-- **`FASTLED_USE_PROGMEM`**: Control PROGMEM usage. Default `0` in `led_sysdefs_esp8266.h`.
+- **`FASTLED_USE_PROGMEM`**: Default `0` in `led_sysdefs_esp8266.h`, but **vestigial on this platform** — it does not control data placement.
+
+  `fastled_progmem.h` dispatches on platform before testing the flag: its `#elif defined(ESP8266)` arm includes `progmem_esp8266.h`, which maps `FL_PROGMEM` to real `PROGMEM` and `FL_PGM_READ_*` to real `pgm_read_*`. The `#if (FASTLED_USE_PROGMEM == 1)` branch lives in an `#else` that ESP8266 never reaches.
+
+  So FastLED tables — palettes, gamma/sine LUTs, noise permutation tables — are flash-resident regardless. A `DEFINE_GRADIENT_PALETTE` links into `.irom0.text` (`0x402xxxxx`), not DRAM.
+
+  Setting it to `1` is a compile error (`platforms/esp/compile_test.hpp`), and would change nothing if it weren't. See [#743](https://github.com/FastLED/FastLED/issues/743).
 - **`FASTLED_ALLOW_INTERRUPTS`**: Allow interrupts during show. Default `1` in `led_sysdefs_esp8266.h`.
 - **`FASTLED_INTERRUPT_RETRY_COUNT`**: Max retries when a frame aborts due to long ISRs/NMIs. Default `2` (see `fastled_config.h`). Used by both single-lane and block drivers.
 

--- a/src/platforms/esp/8266/led_sysdefs_esp8266.h
+++ b/src/platforms/esp/8266/led_sysdefs_esp8266.h
@@ -19,7 +19,21 @@ typedef volatile fl::u32 RwReg;
 typedef fl::u32 prog_uint32_t;
 
 
-// Default to NOT using PROGMEM here
+// Vestigial on ESP8266 -- this value does NOT control data placement here.
+//
+// fastled_progmem.h dispatches on platform before it ever tests this flag:
+// its `#elif defined(ESP8266)` arm pulls in progmem_esp8266.h, which maps
+// FL_PROGMEM to real PROGMEM and FL_PGM_READ_* to real pgm_read_*. The
+// `#if (FASTLED_USE_PROGMEM == 1)` branch sits further down, inside an
+// `#else` that ESP8266 never reaches. FastLED tables are therefore
+// flash-resident on this platform regardless of the 0 below -- verified by
+// symbol placement: a DEFINE_GRADIENT_PALETTE lands in .irom0.text
+// (0x402xxxxx), not DRAM.
+//
+// Kept at 0 because platforms/esp/compile_test.hpp asserts it and user code
+// may branch on it. Do not read it as "ESP8266 keeps constants in RAM" --
+// that was true before the platform dispatch was added, and is exactly what
+// issue #743 reported.
 #ifndef FASTLED_USE_PROGMEM
 # define FASTLED_USE_PROGMEM 0
 #endif

--- a/src/platforms/esp/compile_test.hpp
+++ b/src/platforms/esp/compile_test.hpp
@@ -9,8 +9,14 @@ namespace fl {
 
 #ifdef ESP8266
 void esp8266_compile_tests() {
+// Pinned to 0, but note the flag is inert on ESP8266: fastled_progmem.h
+// selects progmem_esp8266.h from its `#elif defined(ESP8266)` arm, before
+// the `#if (FASTLED_USE_PROGMEM == 1)` test is ever reached. Tables are
+// flash-resident either way. This assert exists so the declared value and
+// the platform's real behavior don't drift further apart -- setting it to 1
+// would not move data, only make the header claim something new. See #743.
 #if FASTLED_USE_PROGMEM != 0
-#error "FASTLED_USE_PROGMEM should be 0 for ESP8266"
+#error "FASTLED_USE_PROGMEM must stay 0 on ESP8266 (it is inert there; PROGMEM is selected by platform dispatch in fastled_progmem.h, not by this flag)"
 #endif
 
 #if !defined(SKETCH_HAS_LARGE_MEMORY_OVERRIDDEN)

